### PR TITLE
Exception messages in FeedBackSessionsLogic class could be extracted to Constants #5357

### LINKED ISSUE #5357: Exception messages in FeedBackSessionsLogic class could be extracted to Constants
There are several EntityDoesNotExistException and InvalidParametersException messages used multiple times in the class. They could all be extracted to Constants.

### PR DESCRIPTION
**Pull request by xeia**
Fixes #5357 

Fixed typo for 'false' which does not seem to have affected the test cases; updated the references to the message string for 'existant' (or should i leave it as it is? running a search for the word 'existant' in returns 24 matches).

### CODE REVIEW

**Comment by YongJieYongJie** [comment]
Just a quick suggestion:
perhaps you can extract a generic error message for `EntityDoesNotExistException` like `"Trying to %s a %s that does not exist"` and use `String.format()` to define the specific individual error message string constants

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 1]
```diff
@@ -138,7 +167,7 @@ public FeedbackSessionAttributes copyFeedbackSession(String newFeedbackSessionNa
 
         if (!coursesLogic.isCoursePresent(courseId)) {
             throw new EntityDoesNotExistException(
-                    "Trying to get feedback sessions for a course that does not exist.");
+                    ERROR_GET_NON_EXISTENT_SESSION_COURSE);
```
Comment (JosephineKwa) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: This could be combined onto one line :)

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 2]
```diff
@@ -263,7 +292,7 @@ public FeedbackSessionQuestionsBundle getFeedbackSessionQuestionsForInstructor(
 
         if (fsa == null) {
             throw new EntityDoesNotExistException(
-                    "Trying to get a feedback session that does not exist.");
+                    ERROR_GET_NON_EXISTENT_SESSION);
```
Comment (JosephineKwa) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: This too

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 3]
```diff
@@ -297,7 +326,7 @@ public FeedbackSessionQuestionsBundle getFeedbackSessionQuestionsForInstructor(
...
```
Comment (JosephineKwa) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: This too. Do also combine the other thrown exceptions in the rest of the changed code below onto one line where applicable (if characters do not exceed the 110 char limit)

**Comment by xeia** [comment]
Updated with both suggestions

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 4]
```diff
@@ -50,6 +50,40 @@
...
+    private static final String ERROR_NUMBER_OF_RESPONSES_EXCEEDS_RANGE = "Number of responses exceeds the limited range";
+    private static final String ERROR_SENDING_EMAILS = "Error while sending emails :";
+    private static final String ERROR_SESSION_UNPUBLISH_ALREADY = "Session is already unpublished.";
+    private static final String ERROR_SESSION_UNPUBLISH_NON_EXISTENT = "Trying to unpublish a non-existent session.";
+    private static final String ERROR_SESSION_UNPUBLISH_PRIVATE = "Private session can't be unpublished.";
+    private static final String ERROR_SESSION_PUBLISH_ALREADY = "Session is already published.";
+    private static final String ERROR_SESSION_PUBLISH_NON_EXISTENT = "Trying to publish a non-existent session.";
+    private static final String ERROR_SESSION_PUBLISH_PRIVATE = "Private session can't be published.";
+    private static final String ERROR_GET_NON_EXISTENT_SESSION_COURSE = "Trying to get feedback sessions for a course that does not exist.";
+    private static final String ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT = "Trying to %s a feedback session %sthat does not exist.";
```
Comment (JosephineKwa) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Maybe we could just stick to using this one and remove the one below, as both strings have roughly the same meaning

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 5]
```diff
@@ -50,6 +50,40 @@
...
+    private static final String ERROR_NON_EXISTENT_SESSION_STRING_FORMAT = "Trying to %s non-existent feedback session%s";
+    private static final String FEEDBACK_SESSION_DOES_NOT_EXIST_CHECK = "check";
```
Comment (JosephineKwa) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: The names for these constants could be changed to `ACTION_<action string>` to show that the type of the string it contains is an action that was performed. i.e. `ACTION_CHECK`, `ACTION_UPDATE`, `ACTION_GET`, `ACTION_REMIND`, `ACTION_VIEW`

**Comment by JosephineKwa** [comment]
@YongJieYongJie , have a look at the changes as well, to see if there are other areas w.r.t the constant formatting that can be changed/improved :)

**Comment by xeia** [comment]
If it is alright, i think the `ERROR_GET_NON_EXISTENT_SESSION_COURSE` can be formatted into `ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT` 
and would changing `"Trying to get a feedback session for student that does not exist."` into `"Trying to get feedback sessions for a student that does not exist.`" be alright?

**Comment by YongJieYongJie** [comment]
Actually rather than using the same `String.format()` whenever the error message is used, you can use `String.format()` to define the specific constants.

For example, rather than using this whenever you need the error message:
~~~
throw new EntityDoesNotExistException(String.format(
                                  ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT,
                                  FEEDBACK_SESSION_DOES_NOT_EXIST_GET, STRING_EMPTY));
~~~

You can define a constant like so:
~~~
ERROR_GETTING_NON_EXISTENT_FEEDBACK_SESSION = String.format(
                                  ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT,
                                  FEEDBACK_SESSION_DOES_NOT_EXIST_GET, STRING_EMPTY);
~~~

and just use the constant:
~~~
throw new EntityDoesNotExistException(ERROR_GETTING_NON_EXISTENT_FEEDBACK_SESSION);
~~~

I think doing so makes the code more readable, but at the same time reduces duplication.

// the same applies for `ACTION_UPDATE` etc.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 6]
```diff
@@ -50,7 +50,47 @@
...
+    private static final String KEY_RANGE = "range";
+    private static final String KEY_SECTION = "section";
+    private static final String KEY_TO_SECTION = "toSection";
+    private static final String KEY_VIEW_TYPE = "viewType";
+    private static final String VALUE_FALSE = "false";
+    private static final String VALUE_TRUE = "true";
+    private static final String STRING_PERIOD = ".";
+    private static final String STRING_COLON = ": ";
+    private static final String ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT = "Trying to %s a feedback session that does not exist%s";
+    private static final String ERROR_SESSION_DOES_NOT_EXIST_EXTRA_PARAM_STRING_FORMAT = "Trying to %s feedback session%s that does not exist.";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: The error here is that the student or course does not exist, so maybe we can rephrase it to something like `"Error getting feedback session(s): %s does not exist."`, where `%s` may be `"student"` or `"course"`

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 7]
```diff
@@ -50,7 +50,47 @@
...
+    private static final String KEY_QUESTION_ID = "questionId";
+    private static final String KEY_RANGE = "range";
+    private static final String KEY_SECTION = "section";
+    private static final String KEY_TO_SECTION = "toSection";
+    private static final String KEY_VIEW_TYPE = "viewType";
+    private static final String VALUE_FALSE = "false";
+    private static final String VALUE_TRUE = "true";
+    private static final String STRING_PERIOD = ".";
+    private static final String STRING_COLON = ": ";
+    private static final String ERROR_SESSION_DOES_NOT_EXIST_STRING_FORMAT = "Trying to %s a feedback session that does not exist%s";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: One of the original problem is that the error messages are not standardized, and trying to extract a constant makes it very obvious.

Maybe we can standardize the error message to always display the `feedbackSessionName` and `courseId`.

So this can be rephrased to something like `"Trying to %s a non-existent feedback session: %s / %s."`, where the three `%s` are the `ACTION`, `feedbackSessionName` and `courseId`.

So we can remove the `STRING_PERIOD` and `STRING_COLON`.

**Comment by xeia** [comment]
@YongJieYongJie That would make it clearer also, would it be alright for me to work on the standardisation/rephrasing in this PR as well? And perhaps in this case
`throw new EntityDoesNotExistException(
                    ERROR_NON_EXISTENT_FS_REMIND
                    + courseId + "/" + feedbackSessionName)`
 the constants, courseId and feedbackSessionName can be extracted into a method for printing? 
Something along the lines of :
`private void String formatExceptionMessage (String action, String courseId, String feedbackSessionName)` that can be used with the formatter String

**Comment by YongJieYongJie** [comment]
@xeia I think rephrasing and standardization that helps in the extraction of constants (like those suggested) is alright.

As for extracting the new formatting method,
I think it can be discussed further in a separate issue because it concerns other classes too. Also I think our error messages as of now are simple enough to be dealt with by `String.format()`.

@JosephineKwa what do you think?

**Comment by xeia** [comment]
@YongJieYongJie Updated with usage of string concatenation. I am thinking of adding a placeholder string format `"%s/%s"` for the `courseId` and `feedbackSessionName` function variables do you think it would be better?

**Comment by JosephineKwa** [comment]
Hey @xeia , just to check, is this ready for review?

**Comment by xeia** [comment]
@JosephineKwa Yeah, just updated and ready for review!

**Comment by wkurniawan07** [comment]
@JosephineKwa let's review this (and other PRs) soon?

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 8]
```diff
@@ -50,7 +50,41 @@
...
+    private static final String KEY_FROM_SECTION = "fromSection";
+    private static final String KEY_IN_SECTION = "inSection";
+    private static final String KEY_IS_INCLUDE_RESPONSE_STATUS = "isIncludeResponseStatus";
+    private static final String KEY_QUESTION_ID = "questionId";
+    private static final String KEY_RANGE = "range";
+    private static final String KEY_SECTION = "section";
+    private static final String KEY_TO_SECTION = "toSection";
+    private static final String KEY_VIEW_TYPE = "viewType";
+    private static final String VALUE_FALSE = "false";
+    private static final String VALUE_TRUE = "true";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: I think for true and false we can keep the string literal in the code without extracting a constant. It is unlikely that we would want to change true to false for everything all at once.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 9]
```diff
@@ -50,7 +50,41 @@
...
+    private static final String ERROR_NON_EXISTENT_FS_STRING_FORMAT = "Trying to %s a non-existent feedback session: %s/%s";
+    private static final String ERROR_NON_EXISTENT_TARGET_STRING_FORMAT = "Error getting feedback session(s): %s does not exist.";
+    private static final String TARGET_STUDENT = "Student";
+    private static final String TARGET_COURSE = "Course";
+    private static final String ERROR_NON_EXISTENT_FS_GET_FOR_COURSE = String.format(ERROR_NON_EXISTENT_TARGET_STRING_FORMAT,
+                                                                                     TARGET_COURSE);
+    private static final String ERROR_NON_EXISTENT_FS_GET_FOR_STUDENT = String.format(ERROR_NON_EXISTENT_TARGET_STRING_FORMAT,
+                                                                                      TARGET_STUDENT);
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: I think can just use
`private static final String ERROR_NON_EXISTENT_FS_GET_FOR_STUDENT = String.format(ERROR_NON_EXISTENT_TARGET_STRING_FORMAT, "Course")`

same for "student", then we can remove `TARGET_STUDENT` and `TARGET_COURSE`

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 10]
```diff
@@ -50,7 +50,41 @@
...
+    private static final String ACTION_GET = "get";
+    private static final String ACTION_UPDATE = "update";
+    private static final String ACTION_CHECK = "check";
+    private static final String ACTION_REMIND = "remind";
+    private static final String ACTION_VIEW = "view";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: I think for each of this, can just use something like:
~~~
private static final String ERROR_GET_NON_EXISTENT_FS =
        String.format(ERROR_NON_EXISTENT_FS_STRING_FORMAT, "get");
~~~

So we can remove all the `ACTION_*`

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 11]
```diff
@@ -50,7 +50,41 @@
...
+    
+    private static final String ASSUMPTION_FAIL_DELETE_INSTRUCTOR = "Fail to delete instructor respondant for ";
+    private static final String ASSUMPTION_FAIL_RESPONSE_ORIGIN = "Client did not indicate the origin of the response(s)";
+    private static final String ERROR_NUMBER_OF_RESPONSES_EXCEEDS_RANGE = "Number of responses exceeds the limited range";
+    private static final String ERROR_SENDING_EMAILS = "Error while sending emails :";
+    private static final String ERROR_SESSION_UNPUBLISH_ALREADY = "Session is already unpublished.";
+    private static final String ERROR_SESSION_UNPUBLISH_NON_EXISTENT = "Trying to unpublish a non-existent session.";
+    private static final String ERROR_SESSION_UNPUBLISH_PRIVATE = "Private session can't be unpublished.";
+    private static final String ERROR_SESSION_PUBLISH_ALREADY = "Session is already published.";
+    private static final String ERROR_SESSION_PUBLISH_NON_EXISTENT = "Trying to publish a non-existent session.";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: This can actually be phrased in the same format as other "Trying to %s a non-existent feedback session: %s/%s" 
~~~
private static final String ERROR_PUBLISH_NON_EXISTENT_FS =
        String.format(ERROR_NON_EXISTENT_FS_STRING_FORMAT, "publish");
~~~

(Also see suggestion below about removing the `ACTION_*`)

**Comment by YongJieYongJie** [comment]
@xeia Added some comments!

**Comment by xeia** [comment]
@YongJieYongJie Noted, I had a dilemma before whether to do a nested string format or string format with string concatenation but went ahead with a one pass string format as i thought it would be better.. Reverted to string concatenation now

Minor: Would it be too confusing for the user to have a single string `"Session is already (un)published."` etc instead of having a separate publish/unpublish string for them?

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 12]
```diff
@@ -50,7 +50,34 @@
...
+    private static final String ERROR_SESSION_PUBLISH_ALREADY = "Session is already published.";
+    private static final String ERROR_SESSION_PUBLISH_PRIVATE = "Private session can't be published.";
+    private static final String KEY_FROM_SECTION = "fromSection";
+    private static final String KEY_IN_SECTION = "inSection";
+    private static final String KEY_IS_INCLUDE_RESPONSE_STATUS = "isIncludeResponseStatus";
+    private static final String KEY_QUESTION_ID = "questionId";
+    private static final String KEY_RANGE = "range";
+    private static final String KEY_SECTION = "section";
+    private static final String KEY_TO_SECTION = "toSection";
+    private static final String KEY_VIEW_TYPE = "viewType";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: I think `KEY_....` should actually be `PARAM_...`

Also, we can group the `PARAM_...` constants either above or below the message-related constants, separated by an empty line.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 13]
```diff
@@ -50,7 +50,34 @@
...
-
+    
+    private static final String ASSUMPTION_FAIL_DELETE_INSTRUCTOR = "Fail to delete instructor respondant for ";
+    private static final String ASSUMPTION_FAIL_RESPONSE_ORIGIN = "Client did not indicate the origin of the response(s)";
+    private static final String ERROR_NUMBER_OF_RESPONSES_EXCEEDS_RANGE = "Number of responses exceeds the limited range";
+    private static final String ERROR_SENDING_EMAILS = "Error while sending emails :";
+    private static final String ERROR_SESSION_UNPUBLISH_ALREADY = "Session is already unpublished.";
+    private static final String ERROR_SESSION_UNPUBLISH_PRIVATE = "Private session can't be unpublished.";
+    private static final String ERROR_SESSION_PUBLISH_ALREADY = "Session is already published.";
+    private static final String ERROR_SESSION_PUBLISH_PRIVATE = "Private session can't be published.";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: To make the error messages consistent as much as possible, maybe these can be along the lines of `"Error publishing feedback session: session is private."`, `"Error publishing feedback session: session is already published."`, `"Error unpublishing feedback session: session is already unpublished."` etc.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 14]
```diff
@@ -50,7 +50,34 @@
 import teammates.storage.api.StudentsDb;
 
 public class FeedbackSessionsLogic {
-
+    
+    private static final String ASSUMPTION_FAIL_DELETE_INSTRUCTOR = "Fail to delete instructor respondant for ";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Let's fix the typo `respondant` => `respondent `

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 15]
```diff
@@ -50,7 +50,34 @@
...
+    private static final String ASSUMPTION_FAIL_RESPONSE_ORIGIN = "Client did not indicate the origin of the response(s)";
+    private static final String ERROR_NUMBER_OF_RESPONSES_EXCEEDS_RANGE = "Number of responses exceeds the limited range";
+    private static final String ERROR_SENDING_EMAILS = "Error while sending emails :";
```
Comment (YongJieYongJie) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Was there a reason for the space before the colon? If not, I think we can remove it.

**Comment by YongJieYongJie** [comment]
@xeia I think it is better to have two separate messages for the published and unpublished.

And the constants look much cleaner now, just left with some rephrasing, correcting typo (and we should be done :+1: ).

**Comment by xeia** [comment]
@YongJieYongJie updated, anyway i made a search for 'respondant' and there are over 200 matches including some of the file names

**Comment by YongJieYongJie** [comment]
@xeia Yup, there are a lot of "respondant" in TEAMMATES, we'll just change the ones our PR touches :+1: 

Anyway, looks good! Tidying constants can be rather trying, but thanks for trying (and doing a good job)!

**Comment by wkurniawan07** [comment]
I will merge this before it gets outdated :p
@xeia once again good job with this PR! :+1:
